[TASK] Running TYPO3 classic mode demo using Docker (#5616)
TYPO3-Documentation/TYPO3CMS-Reference-CoreApi: Documentation/Administration/Installation/ClassicMode/Docker.rst

:navigation-title: Docker demo

..  include:: /Includes.rst.txt
..  _classic-docker-installation:

============================================
Classic TYPO3 demo installation using Docker
============================================

..  note::
    This demo setup is **for local development only** and
    **not suitable for production**.

    It runs TYPO3 in
    `Classic mode <https://docs.typo3.org/permalink/t3coreapi:legacyinstallation>`_
    using the image
    `martinhelmich/typo3 <https://hub.docker.com/r/martinhelmich/typo3>`_
    and a MariaDB container.

    For production or team development, use
    `Composer-based <https://docs.typo3.org/permalink/t3coreapi:installation-composer>`_ TYPO3,
    `Version control (Git) <https://docs.typo3.org/permalink/t3coreapi:version-control>`_

>     and a custom Docker setup.

Author: "and a custom Docker setup." so not the one you see right now. And not supported? That sounds harsh. There is just no ready to use image, which does not mean it is not supported or not possible..

>     and a custom Docker setup.

Author: I added a page stub on using Docker in production

Create a local project directory and subfolders for TYPO3's writable directories:

..  literalinclude:: _codesnippets/_DockerDirDemo.sh
    :language: bash
    :caption: ~/projects/$

..  _classic-docker-network:

2. Create a user-defined Docker network
---------------------------------------

..  code-block:: bash
    :caption: ~/projects/typo3demo/$

    docker network create typo3-demo-net

..  _classic-docker-db:

3. Start the MariaDB database container
---------------------------------------

..  literalinclude:: _codesnippets/_DockerDbDemo.sh
    :language: bash
    :caption: ~/projects/typo3demo/$

..  _classic-docker-typo3:

4. Start the TYPO3 container with mounted writable directories
--------------------------------------------------------------

..  literalinclude:: _codesnippets/_DockerRunTypo3Demo.sh
    :language: bash
    :caption: ~/projects/typo3demo/$

..  important::
    **Ensure folder permissions are correct**

    Mounted folders must be writable by the web server inside the container.
    If you use mounted folders, adjust their permissions on your host system:

    ..  code-block:: bash
        :caption: ~/projects/typo3demo/$

>         chmod -R 777 fileadmin typo3conf typo3temp uploads

Reviewer: This is not recommended for folders. I understand this is the easiest way to solve this so you don't have to fiddle with users and user id. in the container. It might be good to add a warning that this should be avoided to use `777`. 

This is only needed for linux or wsl.

Author: Ok, added a section on how to deal with permissions on different OS, commented it out, left a warning about only to do during development

    :caption: ~/Projects/typo3site/$

    # Stop and remove containers
    docker stop typo3-demo typo3db || true
    docker rm typo3-demo typo3db || true

    # Remove the Docker network
    docker network rm typo3-demo-net || true

    # Remove project folders
    rm -rf fileadmin typo3conf typo3temp uploads

    # Optionally remove Docker images (uncomment if desired)
    # docker rmi martinhelmich/typo3
    # docker rmi mariadb

After this cleanup, you can **repeat the setup instructions** to start fresh
with a clean environment.

..  _classic-docker-command:

Helpful Docker commands
=======================

..  _classic-docker-shell:

Accessing the TYPO3 container shell
-----------------------------------

While the container is running in detached mode, you can open an
interactive shell in the container to inspect files, check logs,
or run TYPO3 console commands.

..  code-block:: bash
    :caption: ~/projects/typo3demo/$

    docker exec -it typo3-demo /bin/bash

This opens an **interactive bash shell** inside the running TYPO3 container.

Type `exit` to leave the container shell.

..  _classic-docker-console:

Running TYPO3 console commands
------------------------------

> TYPO3 provides a **command-line interface (CLI)** via the ``vendor/bin/typo3`` script.

Author: ```suggestion
TYPO3 provides a **command-line interface (CLI)** via the `typo3/sysext/core/bin/typo3` script.
```

    when trying to write configuration, caches, or uploaded files.

..  _classic-docker-access:

5. Access TYPO3 in your browser
-------------------------------

Open:

..  code-block:: text

    http://localhost:8080

Use these database settings:

- **Database Host**: `typo3db`
- **Username**: `db`
- **Password**: `db`
- **Database Name**: `db`

..  _classic-docker-structure:

6. Project directory structure after setup
------------------------------------------

..  directory-tree::

    *   :path:`fileadmin/`
    *   :path:`typo3conf/`
    *   :path:`typo3temp/`

..  note::
    All writable TYPO3 content is now persisted on your local machine.
    TYPO3 Core files are reset when the container stops. You can inspect them
    by `accessing the TYPO3 container shell <https://docs.typo3.org/permalink/t3coreapi:classic-docker-shell>`_.

..  _classic-docker-stop:

7. Stopping and restarting the containers
-----------------------------------------

To stop TYPO3, run:

..  code-block:: bash

    docker stop typo3-demo

To stop and remove the database:

..  code-block:: bash
    :caption: ~/projects/typo3demo/$

    docker stop typo3db

>     docker rm typo3db

Author: ```suggestion
```

:navigation-title: Docker demo

..  include:: /Includes.rst.txt
..  _classic-docker-installation:

============================================
Classic TYPO3 demo installation using Docker
============================================

..  note::
    This demo setup is **for local development only** and
    **not suitable for production**.

    It runs TYPO3 in
    `Classic mode <https://docs.typo3.org/permalink/t3coreapi:legacyinstallation>`_
    using the image
    `martinhelmich/typo3 <https://hub.docker.com/r/martinhelmich/typo3>`_
    and a MariaDB container.

>     For production or team development, use

Reviewer 2: Who is "team development"

Author: when you are developing in a team vs alone?

Reviewer 2: Ups...I have read "our"

    `Composer-based <https://docs.typo3.org/permalink/t3coreapi:installation-composer>`_ TYPO3,
    `Version control (Git) <https://docs.typo3.org/permalink/t3coreapi:version-control>`_
    and a custom Docker setup.

..  contents::

..  _classic-docker-quickstart:

Quick start
===========

To quickly launch TYPO3 in classic mode with Docker:

..  code-block:: bash
    :caption: ~/projects/typo3demo/$

    mkdir -p fileadmin typo3conf typo3temp

    # Ensure TYPO3 can write to these directories
    chmod -R 777 fileadmin typo3conf typo3temp

    docker network create typo3-demo-net

..  literalinclude:: _codesnippets/_DockerDbDemo.sh
    :language: bash
    :caption: ~/projects/typo3demo/$

..  literalinclude:: _codesnippets/_DockerRunTypo3Demo.sh
    :language: bash

>     :caption: ~/projects/typo3demo/$

Reviewer 2: Same caption for two different shell scripts?

Author: both should be run in the same directory

Then open:

..  code-block:: text

    http://localhost:8080

Use these database settings in the TYPO3 installer:

-   **Database Host**: `typo3db`
-   **Username**: `db`
-   **Password**: `db`

> -   **Database Name**: `db`

Reviewer 2: Ahh. These are the default values from Martin Helmich repo. Look like DDEV on first view.

Author: I just used the same default values as DDEV because ppl are used to them. we could also write foo and bar

--------------------------------------------------------------

..  literalinclude:: _codesnippets/_DockerRunTypo3Demo.sh
    :language: bash
    :caption: ~/projects/typo3demo/$

..  important::
    **Ensure folder permissions are correct**

    Mounted folders must be writable by the web server inside the container.
    If you use mounted folders, adjust their permissions on your host system:

    ..  code-block:: bash
        :caption: ~/projects/typo3demo/$

        chmod -R 777 fileadmin typo3conf typo3temp uploads

    Without correct permissions, TYPO3 may fail with **HTTP 500 errors**
    when trying to write configuration, caches, or uploaded files.

..  _classic-docker-access:

5. Access TYPO3 in your browser
-------------------------------

Open:

..  code-block:: text

    http://localhost:8080

Use these database settings:

- **Database Host**: `typo3db`
- **Username**: `db`
- **Password**: `db`
- **Database Name**: `db`

..  _classic-docker-structure:

6. Project directory structure after setup
------------------------------------------

..  directory-tree::

    *   :path:`fileadmin/`
    *   :path:`typo3conf/`
    *   :path:`typo3temp/`

>     *   :path:`uploads/`

Reviewer 2: Is that still there?

>     *   :path:`uploads/`

Author: ```suggestion
```

..  note::
    All writable TYPO3 content is now persisted on your local machine.
    TYPO3 Core files are reset when the container stops. You can inspect them
    by `accessing the TYPO3 container shell <https://docs.typo3.org/permalink/t3coreapi:classic-docker-shell>`_.

..  _classic-docker-stop:

7. Stopping and restarting the containers
-----------------------------------------

To stop TYPO3, run:

..  code-block:: bash

    docker stop typo3-demo

To stop and remove the database:

..  code-block:: bash
    :caption: ~/projects/typo3demo/$

    docker stop typo3db

>     docker rm typo3db

Reviewer 2: In that case also ALL DB data is gone

Author: uh no, we dont want that here, I will remove

    :caption: ~/Projects/typo3site/$

    # Stop and remove containers
    docker stop typo3-demo typo3db || true
    docker rm typo3-demo typo3db || true

    # Remove the Docker network
    docker network rm typo3-demo-net || true

    # Remove project folders
    rm -rf fileadmin typo3conf typo3temp uploads

    # Optionally remove Docker images (uncomment if desired)
    # docker rmi martinhelmich/typo3
    # docker rmi mariadb

After this cleanup, you can **repeat the setup instructions** to start fresh
with a clean environment.

..  _classic-docker-command:

Helpful Docker commands
=======================

..  _classic-docker-shell:

Accessing the TYPO3 container shell
-----------------------------------

While the container is running in detached mode, you can open an
interactive shell in the container to inspect files, check logs,
or run TYPO3 console commands.

..  code-block:: bash
    :caption: ~/projects/typo3demo/$

    docker exec -it typo3-demo /bin/bash

This opens an **interactive bash shell** inside the running TYPO3 container.

Type `exit` to leave the container shell.

..  _classic-docker-console:

Running TYPO3 console commands
------------------------------

> TYPO3 provides a **command-line interface (CLI)** via the ``vendor/bin/typo3`` script.

Reviewer 2: Wrong command. We are in classic mode here ;-)
